File: app/src/main/java/com/saraelmoghazy/base/baselayer/BaseAdapter.java
package com.saraelmoghazy.base.baselayer;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.RecyclerView;

/**
 * Created by Sara Elmoghazy
 */
public abstract class BaseAdapter<T, V extends ViewDataBinding>
        extends RecyclerView.Adapter<BaseViewHolder<T, V>> {

    @NonNull
    @Override
    public BaseViewHolder<T, V> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        V binding = DataBindingUtil.inflate(
                LayoutInflater.from(parent.getContext()), getLayout(), parent, false);
        return new BaseViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull BaseViewHolder<T, V> holder, int position) {
        final T item = getItemForPosition(position);
        holder.bind(item);
    }

    protected abstract T getItemForPosition(int position);

    protected abstract int getLayout();

}

File: app/src/main/java/com/saraelmoghazy/base/baselayer/BaseAdapter.kt
package com.saraelmoghazy.base.baselayer

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView

/**
 * Created by Sara Elmoghazy
 */
abstract class BaseAdapter<T, V : ViewDataBinding> : RecyclerView.Adapter<BaseViewHolder<T, V>>() {

    protected abstract val layout: Int

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder<T, V> {
        val binding = DataBindingUtil.inflate<V>(
            LayoutInflater.from(parent.context), layout, parent, false
        )
        return BaseViewHolder(binding)
    }

    override fun onBindViewHolder(holder: BaseViewHolder<T, V>, position: Int) {
        val item = getItemForPosition(position)
        holder.bind(item)
    }

    protected abstract fun getItemForPosition(position: Int): T

}

